apollo-router/src/plugins/authentication/subgraph_auth.rs

```diff
@@ -0,0 +1,455 @@
+                            err
+                        );
+                            return Ok(ControlFlow::Continue(req));
+                        }
+                    };
+                    // UnsignedPayload only applies to lattice
+                    let signable_request = SignableRequest::new(
+                        req.subgraph_request.method(),
+                        req.subgraph_request.uri(),
+                        req.subgraph_request.headers(),
+                        match signing_params.service_name.as_str() {
+                            "vpc-lattice-svcs" => SignableBody::UnsignedPayload,
+                            _ => SignableBody::Bytes(&body_bytes),
+                        },
+                    );
+
+                    let signing_params = builder.build().expect("all required fields set");
+
+                    let (signing_instructions, _signature) = match sign(signable_request, &signing_params) {
+                        Ok(output) => output,
+                        Err(err) => {
+                            tracing::warn!("Failed to sign GraphQL request for AWS SigV4, skipping signing. Error: {}", err);
+                            return Ok(ControlFlow::Continue(req));
+                        }
+                    }.into_parts();
+                    signing_instructions.apply_to_request(&mut req.subgraph_request);
+                     Ok(ControlFlow::Continue(req))
+                } else {
+                    Ok(ControlFlow::Continue(req))
+                }
+            }
+            }).buffered()
+            .service(service)
+            .boxed()
+        } else {
+            service
+        }
+    }
+}
+
+async fn make_signing_params(config: &AuthConfig) -> SigningParamsConfig {
+    match config {
+        AuthConfig::AWSSigV4(AWSSigV4Config::Hardcoded(config)) => {
+            let region = aws_types::region::Region::new(config.region.clone());
+
+            let chain =
+                aws_config::default_provider::credentials::DefaultCredentialsChain::builder()
+                    .build()
+                    .await;
+            let credentials_provider: Arc<dyn ProvideCredentials> =
+                if let Some(assume_role_provider) = &config.assume_role {
+                    let rp = aws_config::sts::AssumeRoleProvider::builder(
+                        assume_role_provider.role_arn.clone(),
+                    )
+                    .session_name(assume_role_provider.session_name.clone())
+                    .region(region.clone());
+                    let rp = if let Some(external_id) = &assume_role_provider.external_id {
+                        rp.external_id(external_id.as_str())
+                    } else {
+                        rp
+                    };
+
+                    Arc::new(rp.build(chain))
+                } else {
+                    Arc::new(config.clone())
+                };
```

Comment:
This chunk appears to be the same between `Hardcoded` and `DefaultChain`. Could it be extracted into its own function?

Reply:
I gave it a try in 5771292a , but i m not quite satisfied

Reply:
I'm marking the conversation as resolved, feel free to reopen it if you have a better idea ^^